scripts/tuple_tree_generator/tuple_tree_generator/generators/jsonschema.py

```diff
@@ -0,0 +1,144 @@
+
+
+class JSONSchemaGenerator(Generator):
+    def __init__(self, schema, base_namespace: str, root_type):
+        super().__init__(schema, base_namespace)
+        self.root_type = root_type
+
+    def emit(self) -> Dict[str, str]:
+        sources = {
+            "jsonschema.yml": self._emit_jsonschema(self.root_type)
+        }
+        return sources
+
+    def _emit_jsonschema(self, root_type: str):
+        jsonschema = {
+            "$schema": "http://json-schema.org/draft-07/schema#",
+            "$ref": f"#/definitions/{root_type}",
+            "definitions": {
+                "Typeref": {
+                    "type": "string",
+                },
+                "MetaAddress": {
+                    "type": "string",
+                },
+            },
+        }
+        definitions = jsonschema["definitions"]
+
+        for type_definition in self._definitions.values():
+            assert type_definition.name not in definitions
+            if isinstance(type_definition, EnumDefinition):
+                definition = self._enum_jsonschema(type_definition)
+            elif isinstance(type_definition, StructDefinition):
+                definition = self._struct_jsonschema(type_definition)
+            else:
+                raise ValueError()
+
+            definitions[type_definition.tag] = definition
+
+        return yaml.dump(jsonschema)
+
+    def _enum_jsonschema(self, definition: EnumDefinition):
+        members = ["Invalid"]
+        members += [m.name for m in definition.members]
+
+        jsonschema = {
+            "enum": members,
+            "title": definition.tag,
+        }
+        if definition.doc:
+            jsonschema["description"] = definition.doc
+
+        return jsonschema
+
+    def _struct_jsonschema(self, definition: StructDefinition):
+        jsonschema = {
+            "type": "object",
+            "title": definition.tag,
+            "additionalProperties": False,
+        }
+        if definition.doc:
+            jsonschema["description"] = definition.doc
+
+        # TODO: a type can be upcastable even if it is not abstract
+        if definition.abstract:
+            upcastable_to = self._get_upcastable_types(definition)
+            jsonschema["oneOf"] = [
+                {"$ref": f"#/definitions/{f.tag}"} for f in upcastable_to
+            ]
+        else:
+            properties = {}
+
+            for field in definition.all_fields:
+                field_definition = self._convert_struct_field(field)
+                properties[field.name] = field_definition
+
+            jsonschema["required"] = [f.name for f in definition.all_required_fields]
+            jsonschema["properties"] = properties
+
+        return jsonschema
+
+    def _convert_struct_field(self, field: StructField):
+        # TODO: don't hardcode non-standard types
+        revng_string_types = [
+            "Identifier",
+        ]
```

Comment:
Ack for the "not hardcoding" part.
Re: MetaAddress, it's because it gets special treatment for two (related) reasons:
- it's not just any string
- having a separate type for it allows to use a custom (hand-rolled) python class for it, which allows to access the individual components of a MetaAddress easily

Reply:
What does the Python class have to do with jsonschema generation? From a jsonschema POV it's just a string.

Reply:
For instance, in the case of a struct field of type `MetaAddress`, instead of emitting just
```
field:
  type: string
```
I emit
```
field:
  $ref: #/definitions/MetaAddress
```
and a separate definition for MetaAddress which defines it as a string.

Since `MetaAddress` has its own definition, the `datamodel-codegen` creates a `MetaAddress` class (which I can swap with our fancy one) instead of just using a plain string.